java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java

```diff
@@ -139,9 +141,37 @@ private static Tree skipParentheses(Tree parent) {
     return skip;
   }
 
-  private static boolean isSubtype(Type expression, Type target) {
-    return expression.isSubtypeOf(target);
+  private static boolean isUnnecessarySubtypeCast(Type expressionType, TypeCastTree typeCastTree, Type target) {
+    return expressionType.isSubtypeOf(target) && !isLambdaCastNeeded(expressionType, typeCastTree, target);
   }
+
+  private static boolean isLambdaCastNeeded(Type expressionType, TypeCastTree typeCastTree, Type target) {
+    ExpressionTree expression = ExpressionUtils.skipParentheses(typeCastTree.expression());
+    if (expression.is(Tree.Kind.LAMBDA_EXPRESSION)) {
+      List<MethodJavaSymbol> childMethods = getMethodSymbolsOf(expressionType);
+      List<MethodJavaSymbol> parentMethods = getMethodSymbolsOf(target);
+
+      Predicate<JavaSymbol> isDefaultMethod = JavaSymbol::isDefault;
+      Set<MethodJavaSymbol> childDefaultMethods = childMethods.stream().filter(isDefaultMethod).collect(Collectors.toSet());
```

Comment:
This needs to be computed only when the optional is not present and should be done as such in the correct if branch.

Reply:
True, I updated code accordingly